pkg/core/bundle_list.go

```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
 	return b
 }
 
-// ListBundles returns the list of bundle descriptors from a repo.
+// ErrInterrupted signals that the current background processing is interrupted
+var ErrInterrupted = errors.New("background processing interrupted")
+
+// DoSelectBundles is a helper function to listen on a channel of batches of bundle descriptors and an error channel.
+//
+// It applies some function on the received batches and returns an error if one appears on the error channel.
+//
+// Example usage:
+//
+//		err := DoSelectBundles(bundlesChan, errChan, func(bundleBatch model.BundleDescriptors) {
+//			bundles = append(bundles, bundleBatch...)
+//		})
+func DoSelectBundles(bundlesChan <-chan model.BundleDescriptors, errorChan <-chan error, do func(model.BundleDescriptors)) error {
+	var err error
+
+LOOP:
+	// consume batches of ordered bundle metadata
+	for {
+		select {
+		case bundleBatch, ok := <-bundlesChan:
+			if !ok {
+				// last attempt to catch an error
+				// (i.e. bundlesChan has closed, but an error was also present on errorChan and hasn't been selected)
+				if err == nil {
+					if e, ok := <-errorChan; ok {
+						err = e
+					}
+				}
+				break LOOP
+			}
+			do(bundleBatch)
+		case e, ok := <-errorChan: // do not exit here: the only natural exit occurs when bundlesChan is closed
+			if ok && err == nil { // retain first error
+				err = e
+			}
+		}
+	}
+
+	return err
+}
+
+// ApplyBundleFunc is a function to be applied on a bundle
+type ApplyBundleFunc func(model.BundleDescriptor) error
+
+// ListBundlesApply applies some function to the retrieved bundles, in lexicographic order of keys.
 //
-// NOTE: resulting bundles are returned in no particular order.
+// The execution of the applied function does not block background retrieval of more keys and bundle descriptors.
 //
-// TODO: return a paginated list of id<->bd (separate function in repo_list.go)
-func ListBundles(repo string, store storage.Store, opts ...BundleListOption) ([]model.BundleDescriptor, error) {
-	settings := CoreSettings{concurrentBundleList: defaultBundleListConcurrency}
+// Example usage: printing bundle descriptors as they come
+//
+//   err := core.ListBundlesApply(repo, store, func(bundle model.BundleDescriptorOption) error {
+//				fmt.Fprintf(os.Stderr, "%v\n", bundle)
+//				return nil
+//			})
+func ListBundlesApply(repo string, store storage.Store, apply ApplyBundleFunc, opts ...BundleListOption) error {
+	var (
+		err, applyErr error
+		wg            sync.WaitGroup
+		workDone      bool
+		mutex         sync.Mutex
```

Comment:
[the documentation](https://golang.org/pkg/sync/) mentions

> Other than the Once and WaitGroup types, most are intended for use by low-level library routines.

is there a way to communicate other than the shared `workDone` and lock?

Reply:
Mutex is commonplace.

I admit this comes a bit out of the nice full-channel picture... I thought about usIng a signalling channel again... in the end mutex expressed well my intent: first-past-the-post wins

Reply:
Once will do it better.